### ISSUE TITLE
[BUG] Some packages are no longer installable after test command is removed

### ISSUE THREAD

**Issue by chrisirhc** [pypa/setuptools#4519]
For those landing on this issue, please see: 
(thank you @delfick for summarizing this) 
> * This functionality has been deprecated for 5 years, there is a separate issue for discussing if there would have been a better way of removing the functionality [[FR] More gradual breakage of setuptools.command.test #4520](https://github.com/pypa/setuptools/issues/4520)
> * For people using pip (or tools using pip, like poetry), `PIP_CONSTRAINT` set to a file with `setuptools<72.0` should work
> * For people using uv, there seems to be a bug where `UV_CONSTRAINT` doesn't affect the version of setuptools used in build isolation (see [ModuleNotFoundError: No module named 'setuptools.command.test' astral-sh/uv#5551](https://github.com/astral-sh/uv/issues/5551)) and in that case the solution is either `--no-build-isolation` (which should be considered a temporary solution) or getting packages to not use setuptools.command.test or using forks of those packages that comment out the use of setuptools.command.test (until setuptools releases a fix/revert)
> * To press subscribe to the ticket and/or thumbs up the top post instead of adding "same" comments

Quoted from https://github.com/pypa/setuptools/issues/4519#issuecomment-2255389347

---

### setuptools version

setuptools==72.0.0

### Python version

Python 3.9

### OS

Linux

### Description

The breakage change released on 72.0 breaks the default build isolation build of many packages since many of these packages do not pin on a particular setuptools version.
There is also no way to pin to an older setuptools version as pip doesn't offer a way to do this. Installing setuptools==71 first, then installing the package doesn't work as the default build isolation resolves the dependencies without considering the lock file nor currently installed packages.

Based on  these packages need to be patched with `setup_requires` before install, which is nearly impossible since we'd need to patch all packages.

### Expected behavior

Install is successful.

### How to Reproduce

1. Create a virtual environment
2. Upgrade to latest pip 24.2: `python -m pip install --upgrade pip`
3. `pip3.9 install doubles==1.4.0`

### Output

```console
me@my ~/repro-pip-fail
 % pip3.9 install doubles==1.4.0                    
Looking in indexes: http://…
Collecting doubles==1.4.0
  Using cached http://…/doubles-1.4.0.tar.gz (16 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [17 lines of output]
      Traceback (most recent call last):
        File "/home/user/repro-pip-fail/venv/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/home/user/repro-pip-fail/venv/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/home/user/repro-pip-fail/venv/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-h9ppcze_/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 327, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
        File "/tmp/pip-build-env-h9ppcze_/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 297, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-h9ppcze_/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 497, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-h9ppcze_/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 313, in run_setup
          exec(code, locals())
        File "<string>", line 2, in <module>
      ModuleNotFoundError: No module named 'setuptools.command.test'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by databius** [comment]
I got the similar error today.

**Comment by snjtkh** [comment]
we are also getting similar error.

**Comment by nibe12** [comment]
i got a similiar error too

**Comment by jkwe-jhg** [comment]
after an hour of googling, looks like I found the right bug :)

**Comment by vinothsubramanian** [comment]
I am also blocked because of this error 

[https://github.com/IBM/python-sdk-core/issues/202](https://github.com/IBM/python-sdk-core/issues/202)

**Comment by smiledelf** [comment]
after an hour of googling, looks like I found the right bug :)
this is an issue when installing `dbt-core==1.8.4`, which lists `Logbook==1.5.3` as a requirement.

Latest Logbook (1.7.0) seems to have fixed this but dbt-core still points to the older broken module.

**Comment by hashken** [comment]
Facing same error here.

**Comment by popunbom** [comment]
I was also faced this issue with [`cssbeautifier`](https://pypi.org/project/cssbeautifier/) library, and solved by installing with `--no-build-isolation` 
option in my case.

```shell
❯ docker run --rm -it 'python:3.9' /bin/bash

root@81f338564edc:/# python -m pip install --upgrade pip
Requirement already satisfied: pip in /usr/local/lib/python3.9/site-packages (23.0.1)
Collecting pip
  Downloading pip-24.2-py3-none-any.whl (1.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 5.0 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.0.1
    Uninstalling pip-23.0.1:
      Successfully uninstalled pip-23.0.1
Successfully installed pip-24.2
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

root@81f338564edc:/# pip3.9 install cssbeautifier
Collecting cssbeautifier
  Downloading cssbeautifier-1.15.1.tar.gz (25 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [17 lines of output]
      Traceback (most recent call last):
        File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/usr/local/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-sd2vprw8/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 327, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
        File "/tmp/pip-build-env-sd2vprw8/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 297, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-sd2vprw8/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 497, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-sd2vprw8/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 313, in run_setup
          exec(code, locals())
        File "<string>", line 9, in <module>
      ModuleNotFoundError: No module named 'setuptools.command.test'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

root@81f338564edc:/# pip3.9 install --no-build-isolation cssbeautifier
Collecting cssbeautifier
  Using cached cssbeautifier-1.15.1.tar.gz (25 kB)
  Preparing metadata (pyproject.toml) ... done
Collecting jsbeautifier (from cssbeautifier)
  Downloading jsbeautifier-1.15.1.tar.gz (75 kB)
  Preparing metadata (pyproject.toml) ... done
Collecting six>=1.13.0 (from cssbeautifier)
  Downloading six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Collecting editorconfig>=0.12.2 (from cssbeautifier)
  Downloading EditorConfig-0.12.4.tar.gz (13 kB)
  Preparing metadata (pyproject.toml) ... done
Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Building wheels for collected packages: cssbeautifier, editorconfig, jsbeautifier
  Building wheel for cssbeautifier (pyproject.toml) ... done
  Created wheel for cssbeautifier: filename=cssbeautifier-1.15.1-py3-none-any.whl size=30536 sha256=fe16be1043cad7c149be1d60d9320b61f2c64ba672681fab279b2a88a71d9ff4
  Stored in directory: /root/.cache/pip/wheels/7f/2d/19/5d9d82d3c7280e2c68a07b906c458f0a5506cb2b812d330383
  Building wheel for editorconfig (pyproject.toml) ... done
  Created wheel for editorconfig: filename=EditorConfig-0.12.4-py3-none-any.whl size=16377 sha256=9acd52f9cb94f6fabd33cc641d4acb07a4714ada4678388b08e45722eabd5607
  Stored in directory: /root/.cache/pip/wheels/0b/7b/d7/1979a68ec6ce660c89044dc57d175e32f1943bcb8937e43f6e
  Building wheel for jsbeautifier (pyproject.toml) ... done
  Created wheel for jsbeautifier: filename=jsbeautifier-1.15.1-py3-none-any.whl size=94721 sha256=c9fefe5c1d947c492e3b8ea0173783df9821f8ca8a46637a8d4a645b9070ae18
  Stored in directory: /root/.cache/pip/wheels/1f/82/d9/63c09b95f4576d2822b61d073de8f4e3b6fd386112dcb9835d
Successfully built cssbeautifier editorconfig jsbeautifier
Installing collected packages: editorconfig, six, jsbeautifier, cssbeautifier
Successfully installed cssbeautifier-1.15.1 editorconfig-0.12.4 jsbeautifier-1.15.1 six-1.16.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager, possibly rendering your system unusable.It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv. Use the --root-user-action option if you know what you are doing and want to suppress this warning.
```

**Comment by hauntsaninja** [comment]
Here's how to workaround this breakage:

Option 1:
Create a file called constraints.txt
```
setuptools<72
```
Then set `PIP_CONSTRAINT=/path/to/constraints.txt`

Option 2: Use `--no-build-isolation` (and make sure you have an old setuptools installed)

**Comment by Hannes0730** [comment]
> Here's how to workaround this breakage:
> 
> Option 1: Create a file called constraints.txt
> 
> ```
> setuptools<72
> ```
> 
> Then set `PIP_CONSTRAINT=/path/to/constraints.txt` (or `UV_CONSTRAINT` if using uv)
> 
> Option 2: Use `--no-build-isolation` (and make sure you have an old setuptools installed)
> 
> Option 3: Ask the maintainers of each package that is broken like this to adapt to new setuptools and wait for them to make a release
> 
> In general, I find it unfortunate that build isolation intends to increase reproducibility, but in practice the tools given to users make this really tricky in practice. I think ^ is pretty hard for most Python users I know to come up with. Probably wrong place to rant though, setuptools maintainers probably feel the pain of this more than anyone else! :-)

Hi, is option 1 compatible with `poetry add <package-name>`? or is that pip dependent? I tried option 2 but I think `--no-build-isolation` is not supported in `poetry (1.8.2)`. 

`RUN poetry install --no-interaction --no-build-isolation --without dev --no-ansi -vvv`

**Comment by anyidea** [comment]
Same issue.

**Comment by Luke31** [comment]
Would it be possible to mark this version as yanked in pypi and follow a more gradual approach with deprecation as mentioned in https://github.com/pypa/setuptools/issues/4520 ?

Especially installs with poetry seem to have issues and currently I couldn't find a solution to overcome it. Did anybody find a solution for following or similar combination?
- docker
- poetry==1.7.1
- python:3.10

(Of course we want to upgrade, but we're still using sqlalchemy 1.4.49 which requires an major upgrade to 2.0 to support  PEP 517).

**Comment by melvin806** [comment]
> Here's how to work around this breakage:
> 
> Option 1: Create a file called constraints.txt
> 
> ```
> setuptools<72
> ```
> 
> Then set `PIP_CONSTRAINT=/path/to/constraints.txt`
> 
> Option 2: Use `--no-build-isolation` (and make sure you have an old setuptools installed)
> 
> Option 3: Ask the maintainers of each package that is broken like this to adapt to new setuptools and wait for them to make a release
> 
> In general, I find it unfortunate that build isolation intends to increase reproducibility, but the tools given to users make this really tricky in practice. I think ^ is pretty hard for most Python users I know to come up with (and lacks e.g. the ability to hash pin). Probably wrong place to rant though, setuptools maintainers probably feel the pain of this more than anyone else! :-)
> 
> Edit: if setuptools does decide to revert something here, I have a suggestion in #4520 for how the underlying change could be rolled out more gradually

may i know how do we apply one of these possible workaround for our CICD pipeline. its totally broken and we cannot build a new build image.

**Comment by Hannes0730** [comment]
> > Here's how to workaround this breakage:
> > Option 1: Create a file called constraints.txt
> > ```
> > setuptools<72
> > ```
> > Then set `PIP_CONSTRAINT=/path/to/constraints.txt` (or `UV_CONSTRAINT` if using uv)


I tried doing `Option 1` in the dockerfile, but the error persists. 
```
FROM python:3.12.2-slim as python
ENV PYTHONUNBUFFERED=true
WORKDIR /app


FROM python as poetry
ENV POETRY_HOME=/opt/poetry
ENV POETRY_VIRTUALENVS_IN_PROJECT=true
ENV PATH="$POETRY_HOME/bin:$PATH"
RUN python -c 'from urllib.request import urlopen; print(urlopen("https://install.python-poetry.org").read().decode())' | python -
COPY . ./
ENV PIP_CONSTRAINT=/app/constraints.txt
RUN poetry install --no-interaction --without dev --no-ansi -vvv ---> ERROR: Note: This error originates from the build backend, and is likely not a problem with poetry but with jsbeautifier (1.15.1) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "jsbeautifier (==1.15.1)"
```

Am I missing something? I tried deleting the docker volume so that it installs all dependencies from scratch.

**Comment by kndzuedemir** [comment]
We're also facing this issue today. One of our package `channels[daphne]==4.0.0` depends `setuptools` and we got errors below during build;

```
31.15         File "/tmp/pip-build-env-wyuei3de/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 313, in run_setup
31.15           exec(code, locals())
31.15         File "<string>", line 21, in <module>
31.15       NameError: name 'TestCommand' is not defined
```

**Comment by ilpianista** [comment]
> Especially installs with poetry seem to have issues and currently I couldn't find a solution to overcome it. Did anybody find a solution for following or similar combination?

After several attempts, it looks like this one worked:
```
$ pip install <problematic dependecy>
$ poetry run pip install <problematic dependecy>
$ poetry install
```

**Comment by solarissmoke** [comment]
This completely broke Poetry installation for us, because Poetry enforces build isolation, so you cannot just install your preferred version of setuptools. Ended up with this workaround - exporting the poetry lock file to `requirements.txt` and then installing with pip:

```
poetry export -f requirements.txt -o requirements.txt --no-ansi -n --without-hashes
pip install setuptools<72
pip install -r requirements.txt
```

**Comment by elliexcoding** [comment]
I tried using `--no-build-isolation` but my Docker build will fail as it will post `ModuleNotFoundError: No module named 'poetry'` from dist-packages.

**Comment by sandykr01** [comment]
Good morning, everyone! Yup, ran into a similar problem when our GH workflows started crashing after setuptools dropped this major update. For me, it’s breaking down while installing SQLAlchemy:
```
Collecting SQLAlchemy==1.3.24 (from -r requirements.txt (line 22))
  Downloading SQLAlchemy-1.3.24.tar.gz (6.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.4/6.4 MB 119.3 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'error'
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      Traceback (most recent call last):
        File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-uiic32ek/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 327, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-uiic32ek/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 297, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-uiic32ek/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 497, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-uiic32ek/overlay/lib/python3.11/site-packages/setuptools/build_meta.py", line 313, in run_setup
          exec(code, locals())
        File "<string>", line 14, in <module>
      ModuleNotFoundError: No module named 'setuptools.command.test'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error
```

**Comment by Artui** [comment]
Issue for us as well. Persists even though setuptools are pinned to 71.1.0.
<img width="1665" alt="image" src="https://github.com/user-attachments/assets/96eab7cd-da8a-4fdc-85ed-3d3f3f86f6c3">

**Comment by elliexcoding** [comment]
> Issue for us as well. Persists even though setuptools are pinned to 71.1.0. 

Have you tried running a pip install via requirements.txt rather than poetry.lock/pyproject.toml?  That was what got it working for me.

**Comment by ilpianista** [comment]
> Issue for us as well. Persists even though setuptools are pinned to 71.1.0.

@Artui have a look to https://github.com/pypa/setuptools/issues/4519#issuecomment-2255263220

**Comment by hashken** [comment]
For our pex builds, the pip constraint approach outlined in https://github.com/pypa/setuptools/issues/4519#issuecomment-2254983472 solves the issue

```
echo "setuptools<72" > /tmp/constraint.txt
EXPORT PIP_CONSTRAINT = /tmp/constraint.txt
<run the pex command>
```

**Comment by agusmakmun** [comment]
In my case, we disable some packages that causing this issue, for example:

```
# elastic-apm==6.6.0  # https://github.com/elastic/apm-agent-python
# djlint==1.34.1  # https://github.com/Riverside-Healthcare/djLint
```

**Comment by AmosRosmarin** [comment]
`pip install -I --force-reinstall setuptools==71.0.0`

works for me

**Comment by simonwhitaker** [comment]
Here's a repo with some Github Actions jobs demonstrating how you can work around this in your automated workflows.

https://github.com/simonwhitaker/setuptools-demo

**Comment by Artui** [comment]
> > Issue for us as well. Persists even though setuptools are pinned to 71.1.0.
> 
> @Artui have a look to [#4519 (comment)](https://github.com/pypa/setuptools/issues/4519#issuecomment-2255263220)

requirements approach works, but we run poetry everywhere, would be nice to have a workaround for that too.

Was that command deprecated first at some point? Or was it removed right away? Feels like the release should mark it as deprecated first instead of ripping it out...

**Comment by Hannes0730** [comment]
> Quote reply

**FOR POETRY**
I think this worked for me with PIP_CONSTRAINT combined. 

```
FROM python:3.12.2-slim as python
ENV PYTHONUNBUFFERED=true
WORKDIR /app

FROM python as poetry
ENV POETRY_HOME=/opt/poetry
ENV POETRY_VIRTUALENVS_IN_PROJECT=true
ENV PATH="$POETRY_HOME/bin:$PATH"
RUN python -c 'from urllib.request import urlopen; print(urlopen("https://install.python-poetry.org").read().decode())' | python -
COPY . ./

# Install litestar because of broken setuptools unable to install jsbeautifier
ENV PIP_CONSTRAINT=/app/constraints.txt
RUN pip install litestar["standard"]
RUN poetry run pip install litestar["standard"]


RUN poetry install --no-interaction --without dev --no-ansi -vvv
```

**Comment by AfikGrinstein** [comment]
Can we revert this PR https://github.com/pypa/setuptools/pull/4458 ?
I tried everything and was still not able to get it to work.
We are working with poetry.

**Comment by bissli82** [comment]
Same here - please revert :(

**Comment by illia-khorunzhyi-metro** [comment]
With this work for me with Docker, python 3.12

```
RUN pip3 install -I --force-reinstall setuptools==71.0.0
RUN pip3 install --no-build-isolation --no-cache-dir -r requirements.txt
```

**Comment by AfikGrinstein** [comment]
> > [AfikGrinstein](/AfikGrinstein)
> 
> Hi, have you tried [#4519 (comment)](https://github.com/pypa/setuptools/issues/4519#issuecomment-2255337022) ? I also use poetry and it worked.

Yes.
Didn't work for me. :-(

**Comment by petrprikryl** [comment]
> Same here - please revert :(

+1

**Comment by samzong** [comment]
Same issue with me.

**Comment by apakhomov** [comment]
Revert plz, this breaks lots of projects

**Comment by bjgrosse** [comment]
This worked for us using poetry and docker. In the dockerfile:

```
WORKDIR /code
# these two new lines are the fix
RUN echo "setuptools<72" > "constraints.txt"
ENV PIP_CONSTRAINT=/code/constraints.txt

RUN poetry install --no-interaction --no-ansi
```

**Comment by wfgydbu** [comment]
same here +1, with Nylas sdk

**Comment by 0noDERA** [comment]
I got a similar error.

**Comment by delfick** [comment]
@chrisirhc  are you able to please add to the description of the ticket that 

* This functionality has been deprecated for 5 years, there is a separate issue for discussing if there would have been a better way of removing the functionality https://github.com/pypa/setuptools/issues/4520
* For people using pip, `PIP_CONSTRAINT` set to a file with `setuptools<72.0` should work
* For people using uv, there seems to be a bug where `UV_CONSTRAINT` doesn't affect the version of setuptools (see https://github.com/astral-sh/uv/issues/5551) and in that case the solution is either `--no-build-isolation` (which should be considered a temporary solution) or getting packages to not use setuptools.command.test or using forks of packages that do
* To press subscribe to the ticket or thumbs up the original instead of adding "same" comments

:)

**Comment by Luke31** [comment]
@ilpianista @Hannes0730 Thanks for your help, following worked for me too in poetry:

```
RUN echo "setuptools<72" > /code/constraints.txt
ENV PIP_CONSTRAINT=/code/constraints.txt
RUN poetry run pip install sqlalchemy==1.4.49
RUN poetry install
```
(Somehow `RUN pip install sqlalchemy==1.4.49` wasn't required but that might be due to using python 3.10 🤔

**Comment by halbow** [comment]
FYI for people using pip-tools, installing setuptools version 71.1.0 and make sure you use pip-sync and pip-compile with the `--pip-args --no-build-isolation` flag

```
pip install --upgrade pip-tools pip wheel;
pip install setuptools==71.1.0;
pip-sync requirements.txt --pip-args --no-build-isolation
```

**Comment by Artui** [comment]
> This worked for us using poetry and docker. In the dockerfile:
> 
> ```
> WORKDIR /code
> # these two new lines are the fix
> RUN echo "setuptools<72" > "constraints.txt"
> ENV PIP_CONSTRAINT=/code/constraints.txt
> 
> RUN poetry install --no-interaction --no-ansi
> ```

Doesn't seem to work for us... I've no idea what's going on "under the hood" here, but it seems to me that poetry is not respecting the setuptools of the virtualenv (we're using `python:3.12.4-slim` here)
<img width="1565" alt="image" src="https://github.com/user-attachments/assets/04761224-e28d-424f-8e45-c376b8b41e78">

**Comment by stamparm** [comment]
seems that this one is going out of the control, based on quick Github [search](https://github.com/search?q=ModuleNotFoundError+-+No+module+named+%27setuptools.command.test%27&type=issues&s=comments&o=desc). it would be really nice if somebody from the dev team would make a definite statement whether we should do "dirty patches" in our deployment scripts with some random workarounds OR we should wait for the resolution in the `setuptools` itself. in our case, we have more than a dozen of projects with custom deployment scripts and complex inter-dependencies.

currently closely watching for this issue/thread to decide on future steps

**Comment by CesarFalco** [comment]
same here +1

**Comment by bissli82** [comment]
> @ilpianista @Hannes0730 Thanks for your help, following worked for me too in poetry:
> 
> ```
> RUN echo "setuptools<72" > /code/constraints.txt
> ENV PIP_CONSTRAINT=/code/constraints.txt
> RUN poetry run pip install sqlalchemy==1.4.49
> RUN poetry install
> ```
> 
> The current workaround for potery seems to be... not using poetry to install 👌 (Somehow `RUN pip install sqlalchemy==1.4.49` wasn't required but that might be due to still using python 3.10 🤔 )

This workaround seems to be working, install the problematic package via pip(with constraints) and the rest via poetry

**Comment by eyareyar123** [comment]
Same issue here +1

**Comment by nakamichiworks** [comment]
For poetry users, exporting `requirements.txt` and `pip install` with the constraint worked for me.

```
export PIP_CONSTRAINT=constraints.txt
echo "setuptools<72" > constraints.txt
poetry export -o requirements.txt
pip install -r requirements.txt
```

**Comment by ferrarijefferson** [comment]
For `uv` users, this worked for me:

```shell
uv venv
source .venv/bin/activate
uv pip install --upgrade pip-tools pip wheel
uv pip install setuptools==71.1.0 poetry==1.8.3
 uv pip install --no-build-isolation -r ../requirements-dev.txt
```

**Comment by jamesbuddrige** [comment]
Also unable to use poetry at the moment...

**Comment by pfaion** [comment]
This breaks ~10 packages in my dependency chain which just specify:
```
[build-system]
requires = ["setuptools"]
```

so they don't pin their setuptools build dependency, which makes them pull the latest version on every build. And that build happens in the isolated build environments for each package, so I don't even really directly affect that somehow.

Not a good start into the week.

Makes me wonder if you should generally pin your build environment, there shouldn't be any harm since it does not conflict with anything else? Assuming that you _use_ build isolation though.

**Comment by omerbe-nym** [comment]
> #4458

Thank you, this solved our issue (using Poetry 1.8.3, Python 3.11.4).

```
RUN echo "setuptools<72" > constraints.txt

ENV PIP_CONSTRAINT=constraints.txt

RUN poetry config virtualenvs.create false && \
    poetry run pip install sqlalchemy==1.3.24 && \
    poetry install --no-cache && \
```

I could not understand why when I degraded the setuptools to version below 72.0.0 it kept cruising with the error of `ModuleNotFoundError: No module named 'setuptools.command.test'`.

**Comment by apakhomov** [comment]
Guys, advising "downgrade setuptools, downgrade pip" and putting minuses under revert proposals, it's easy to do when you develop single project on your own pc in your own venv. When you have tens or hundreds of projects, that use dependencies of dependencies, the game becomes much more exciting. rebuild base images, rebuild depending images, rebuild libraries... Do you really understand the scope of work?

**Comment by amine-el-amrani** [comment]
Same issue here +1

**Comment by younsihamza** [comment]
I was facing this problem when I wanted to install Django Channels with Daphne. I found this solution and used the following command:
`pip3 install --no-build-isolation --no-cache-dir "setuptools<72"`

**Comment by solebox** [comment]
use this in our docker images:
``` 
RUN echo "setuptools<72" > /app/constraints.txt
ENV PIP_CONSTRAINT=/app/constraints.txt
```

**Comment by solebox** [comment]
thanks @omerbe-nym

**Comment by justinclift** [comment]
@jamesbuddrige Try the stuff in https://github.com/pypa/setuptools/issues/4519#issuecomment-2255446798/

**Comment by andrewokello** [comment]
For those using docker and poetry, 

This is how my dockerfile looks now, and it works for me

`# ---- Base Python ----
FROM python:3.11.4-slim AS base

# Create app directory
WORKDIR /app

# ---- Dependencies ----
FROM base AS dependencies

# Install pip and Poetry
RUN pip install --upgrade pip && \
    pip install poetry

# Copy poetry.lock* in case it doesn't exist in the repo
COPY ./poetry.lock ./pyproject.toml /app/

# Install project dependencies.
RUN export PIP_CONSTRAINT=constraints.txt
RUN echo "setuptools<72" > constraints.txt
RUN poetry export -o requirements.txt
RUN pip install setuptools==71.1.0
RUN pip install --no-build-isolation -r requirements.txt

# ---- Copy Files/Build ----
FROM dependencies AS build

WORKDIR /app

COPY . /app

# --- Release ----
FROM base AS release

COPY --from=dependencies /usr/local/bin /usr/local/bin
COPY --from=dependencies /usr/local/lib/python3.11/site-packages /usr/local/lib/python3.11/site-packages
COPY --from=build /app /app

EXPOSE 8000

# Setup entrypoint
COPY ./docker/entrypoint.sh /app

ENTRYPOINT ["sh", "/app/entrypoint.sh"]
CMD ["gunicorn", "config.wsgi:application", "--bind", "0.0.0.0:8000"]`

**Comment by LewisCowlesMotive** [comment]
It'd be great to know when this is solved so folks can upgrade again. `setuptools` wasn't even an explicit dependency and now it is an explicit devdependency.

**Comment by berkaycamur** [comment]
Guys I'm facing with an issue during installation with poetry on docker and in first it says `ModuleNotFoundError: No module named 'setuptools.command.test'` and then it returns error about a package which is trying to be installed. Is that problem also depends on that error?

**Comment by omerbe-nym** [comment]
> Guys I'm facing with an issue during installation with poetry on docker and in first it says `ModuleNotFoundError: No module named 'setuptools.command.test'` and then it returns error about a package which is trying to be installed. Is that problem also depends on that error?

https://github.com/pypa/setuptools/issues/4519#issuecomment-2255446798

**Comment by jamesbuddrige** [comment]
> @jamesbuddrige Try the stuff in [#4519 (comment)](https://github.com/pypa/setuptools/issues/4519#issuecomment-2255446798)

We're working across 50+ repo's so this isn't an ideal solution, but will give it a go.

**Comment by agusmakmun** [comment]
@jaraco can you please roll back the latest version?  As it impacting many packages.
After that, we can do the deprecate solution to ensure old packages still supported.

**Comment by petrprikryl** [comment]
Could it 72.0.0 marked as yanked before other essential tools will be working with this change? :pray:

**Comment by llybin** [comment]
pipenv

Pipfile
```
[packages]
setuptools = "==71.1.0"  # sse docker/Dockerfile:54
```

pipenv lock

Dockerfile
```
# https://github.com/pypa/setuptools/issues/4519#issuecomment-2255390400
# try to remove these lines after the issue is resolved
# remove setuptools from Pipfile
# also perhaps need wait for when other packages will be updated with compatible setuptools
RUN echo "setuptools<72" > /app/constraints.txt
ENV PIP_CONSTRAINT=/app/constraints.txt
RUN pip install --no-build-isolation --no-cache-dir "setuptools<72"
```

**Comment by gmyrianthous** [comment]
Yank `v72.0.0` ASAP!

**Comment by royire** [comment]
Yank, Yank, Yank!!!

**Comment by elliottrabac** [comment]
> > after an hour of googling, looks like I found the right bug :) this is an issue when installing `dbt-core==1.8.4`, which lists `Logbook==1.5.3` as a requirement.
> > Latest Logbook (1.7.0) seems to have fixed this but dbt-core still points to the older broken module.
> 
> For some reason, this seems to work fine now with us:
> 
> `python -m pip install --force-reinstall setuptools==71.1.0 && python -m pip install --no-cache-dir --no-build-isolation dbt-snowflake==1.7.2`
> 
> I don't know man

Works for us as well to fix the dbt package issue 👍

**Comment by djuarezg** [comment]
Did someone figure out how to fix this on Pants package builds?

**Comment by berkaycamur** [comment]
> Did someone figure out how to fix this on Pants package builds?
firstly apply these and then run the standart installing command for the spesific package #4519 (comment)

**Comment by tisnik** [comment]
For anybody using PDM, there's our solution - not nice, but we have the need to pass our CI ASAP:

```
        # force install older setuptools version
	pip install -I --force-reinstall setuptools==71.0.0
	pip show setuptools
	export PIP_CONSTRAINT=constraints.txt
	pdm install --no-isolation --dev -v
```

(`--dev` is not needed if you want to install non-dev dependencies, and `-v` is just to increase verbosity level)

**Comment by berkaycamur** [comment]
Guys in brief, the main solution is downgrade the version of the setuptools ( 69.0.2 worked fine on me ). There is bunch of solutions above but in my case, setuptools was preventing me to install a package using poetry and  #4519 (comment) worked for me

**Comment by villoro** [comment]
I also had problems with this. In my case I need to install the packages in a github action and in docker with poetry. It's a DBT project, so as mentioned in https://github.com/pypa/setuptools/issues/4519#issuecomment-2254942140 it's because of `logbook`. So here is how I fixed it

# 1. Constraints
Create the `constraints.txt` with just `setuptools<72`

# 2. Docker
```docker
COPY constraints.txt .
ENV PIP_CONSTRAINT=constraints.txt

RUN poetry run pip install logbook==1.5.3 && \
    poetry install --no-cache
```

> Replace `logbook==1.5.3` with whatever package is giving you problems

# 3. Github action
```yaml
    - name: Handle nasty bug # https://github.com/pypa/setuptools/issues/4519#issuecomment-2255390400
      run: |
        echo "PIP_CONSTRAINT=constraints.txt" >> $GITHUB_ENV
        poetry run pip install logbook==1.5.3

    - name: Install dependencies
      run: poetry install --no-cache

    # I had some problems with `pre-commit`, so I had to remove the env var afterwards:
    - name: Remove PIP_CONSTRAINT
      run: echo "PIP_CONSTRAINT=" >> $GITHUB_ENV
```

> Replace `logbook==1.5.3` with whatever package is giving you problems

Hope it helps!

**Comment by RubTalha** [comment]
https://stackoverflow.com/a/78806669/26556668

**Comment by chirag-jn** [comment]
> I also had problems with this. In my case I need to install the packages in a github action and in docker with poetry. It's a DBT project, so as mentioned in [#4519 (comment)](https://github.com/pypa/setuptools/issues/4519#issuecomment-2254942140) it's because of `logbook`. So here is how I fixed it
> 
> # 1. Constraints
> Create the `constraints.txt` with just `setuptools<72`
> 
> # 2. Docker
> ```
> COPY constraints.txt .
> ENV PIP_CONSTRAINT=constraints.txt
> 
> RUN poetry run pip install logbook==1.5.3 && \
>     poetry install --no-cache
> ```
> 
>  > Replace `logbook==1.5.3` with whatever package is giving you problems
> 
> # 3. Github action
> ```yaml
>     - name: Handle nasty bug # https://github.com/pypa/setuptools/issues/4519#issuecomment-2255390400
>       run: |
>         echo "PIP_CONSTRAINT=constraints.txt" >> $GITHUB_ENV
>         poetry run pip install logbook==1.5.3
> 
>     - name: Install dependencies
>       run: poetry install --no-cache
> 
>     # I had some problems with `pre-commit`, so I had to remove the env var afterwards:
>     - name: Remove PIP_CONSTRAINT
>       run: echo "PIP_CONSTRAINT=" >> $GITHUB_ENV
> ```
> 
>  > Replace `logbook==1.5.3` with whatever package is giving you problems
> 
> Hope it helps!

This solution worked for me!

**Comment by LewisCowlesMotive** [comment]
https://github.com/pypa/setuptools/issues/4519#issuecomment-2254983472 does not work for folks who use poetry.

**Comment by Afoucaul** [comment]
For Poetry/Docker users, I've managed to contain the fix to these three steps, though I did have to iterate to find all the dependencies that were broken

```Dockerfile
RUN mkdir -p /opt/pip && echo "setuptools<72" >/opt/pip/constraints.txt
ENV PIP_CONSTRAINT=/opt/pip/constraints.txt
RUN \
  mkdir /build/ \
  && poetry export --without-hashes \
    | grep \
      -e intervaltree \
      -e vcrpy \
      # -e any_other_package
    >/build/requirements.setuptools-bugged.txt \
  && poetry run pip install -r /build/requirements.setuptools-bugged.txt
```

**Comment by pcreux** [comment]
![dependency_xkcd_setuptools](https://github.com/user-attachments/assets/6b1be885-d128-4dd3-8247-03ec9e7cdde4)

**Comment by mwPandoraid** [comment]
please yank this, none of these workarounds should be necessary and some of them don't even work - disabling build isolation results in build errors for me.

i understand the desire to remove something that's obsolete but clearly this wasn't the right moment or way to do it.

**Comment by deceze** [comment]
This worked for me for Poetry without picking out individual broken packages:

```
poetry export -o requirements.txt --without-hashes
poetry export -f constraints.txt -o constraints.txt --without-hashes
PIP_CONSTRAINT="$(pwd)/constraints.txt" poetry run pip -r requirements.txt
```

In other words, let `poetry` export requirements and constraints to files, and make sure you use the same options as you'd use for `poetry install` (e.g. `--only main`). You may or may not need `--without-hashes`, depending on where you're getting what dependencies from. Try omitting `--without-hashes`, it's probably better to use hashes if it works for you. Finally let `pip` install the dependencies from `requirements.txt` and obeying the constraints. This assumes you do have `setuptools<72` locked in Poetry, otherwise the constraints won't do much.

**Comment by kakuchotakaku** [comment]
I was also able to work around it using this method. https://github.com/pypa/setuptools/issues/4519#issuecomment-2254983472
Just to leave a note for reference, setting `virtualenvs.create false` is not mandatory.

**Comment by gmyrianthous** [comment]
Any workarounds for `poetry` ?

```
setuptools = "<72.0"
```
doesn't seem to work for some reason.

**Comment by JohnOMDev** [comment]
> after an hour of googling, looks like I found the right bug :) this is an issue when installing `dbt-core==1.8.4`, which lists `Logbook==1.5.3` as a requirement.
> 
> Latest Logbook (1.7.0) seems to have fixed this but dbt-core still points to the older broken module.

Yeah. Exact same issue. `dbt-core==1.8.3` also pointed to Logbook==1.5.3. I tried to install Logbook==1.7.0 seperately but had conflict issue. It is failing only my C1/CD and I believed dbt needed to fix it asap.

**Comment by miandreu** [comment]
The version 72.0.0 is currently not functioning correctly. Please YANK!!!!! We can implement five different workarounds for various use cases, but the solution shouldn´t be this one. YANK!!

**Comment by manugarri** [comment]
this is ridiculous, this package is one of the most core packages and it has over 12 years of runway.   Im very surprised as this bug is obvious the moment a realisitic test build is perform.

Please yank this version.

**Comment by wyd3x** [comment]
Pls YANK it!! All my CI dead bcs of it

Crowdstrike round #2?

**Comment by delfick** [comment]
@chrisirhc we may wanna add a couple more dotpoints

* the problem is with the package manager building wheels in an [isolated environment](https://peps.python.org/pep-0518/) and installing it's own version of setuptools for this purpose (which is why the version specified in your environment doesn't make a difference). The package manager ends up using the latest version of setuptools which doesn't have a feature used by some packages at build time. The `PIP_CONSTRAINT` environment variable mentioned below constrains what versions the package manager uses for this purpose
* Yanking this release is a popular request, the devs have yet to respond to this issue with what will happen moving forward. The solutions in this thread are only meant as temporary solutions until a real fix happens in setuptools.

**Comment by alpoi-x** [comment]
Maybe unpopular opinion - don't yank it! The test command has been [deprecated](https://github.com/pypa/setuptools/issues/1684) for _5 years_, which is more than enough time for people to update their packages. Clearly something drastic like this was necessary to force stragglers to make the change (else it wouldn't be an issue).

If the packages you are using don't resolve this in the near future, maybe consider not using unmaintained packages ¯\_(ツ)_/¯

**Comment by Artui** [comment]
> Maybe unpopular opinion - don't yank it! The test command has been [deprecated](https://github.com/pypa/setuptools/issues/1684) for _5 years_, which is more than enough time for people to update their packages. Clearly something drastic like this was necessary to force stragglers to make the change (else it wouldn't be an issue).
> 
> If the packages you are using don't resolve this in the near future, maybe consider not using unmaintained packages ¯_(ツ)_/¯

While I agree with this in generally, [this issue](https://github.com/pypa/setuptools/issues/4520) explains the problem with this "deprecation" quite well? As in - it should've been at least "screaming" in the console each time the deprecated piece of code was utilised. That way consumers of packages with this deprecation could've reached out to maintainers, asking for upgrade or smth similar

**Comment by 712u3** [comment]
my poetry generates bad links for some reason, so fast and ugly workaround for today is

```python
ENV PIP_CONSTRAINT=constraints.txt
RUN echo "setuptools<72" > constraints.txt && \
    poetry export --without-hashes -o requirements.txt && \
    sed -i 's/git+git+ssh/git+ssh/g' requirements.txt
RUN --mount=type=ssh pip install -r requirements.txt
```

seems that i should have also used `--no-build-isolation` but everything works well without it

p.s. don't yank) let's finally update things

**Comment by mwPandoraid** [comment]
> my poetry generates bad links for some reason, so fast and ugly workaround for today is
> 
> ```python
> ENV PIP_CONSTRAINT=constraints.txt
> RUN echo "setuptools<72" > constraints.txt && \
>     poetry export --without-hashes -o requirements.txt && \
>     sed -i 's/git+git+ssh/git+ssh/g' requirements.txt
> RUN --mount=type=ssh pip install -r requirements.txt
> ```
> 
> seems that i should have also used `--no-build-isolation` but everything works well without it
> 
> p.s. don't yank) let's finally update things

holding prod environments hostage is not the best way to go about making package maintainers do their job is my opinion on the matter

**Comment by KPrasch** [comment]
I am having this issue with the `autobahn` package while installing with pip (https://github.com/crossbario/autobahn-python/issues/1644).   

Changing my local projects requirements was not enough - the new version of setuptools was still being used despite pinning setuptools<72 for the reasons mentioned here https://github.com/pypa/setuptools/issues/4519#issuecomment-2255690881

Since I'm using poetry to build pip `requirements.txt`, I was able to successfully use `constraints.txt` as mentioned above (https://github.com/pypa/setuptools/issues/4519#issuecomment-2255652133).  


Thanks for the help everyone! 😅

**Comment by sstegmueller** [comment]
FWIW exporting the `PIP_CONSTRAINT=path/to/constraints.txt` variable works as a workaround for pipenv too.

**Comment by stamparm** [comment]
not sure how to write this down without offending anyone. pretty much similar behavior as with python core devs - "[4 letter word] you, you've been warned in time, not our fault". some people just want to see the world burn

p.s. at the end we decided to use the dirty hack with `PIP_CONSTRAINT` in our production environment until the dust settles

**Comment by kalomaze** [comment]
![community-donald-glover](https://github.com/user-attachments/assets/4a6bf3fc-967e-4f81-bff8-6f2d24afb198)

**Comment by noiji** [comment]
for those using poetry, this helped solve the issue.

```
echo "setuptools<72" > constraints.txt
export PIP_CONSTRAINT=constraints.txt
poetry run pip install ffmpy==0.3.2 #  package with the error
poetry install
```

but still dunno how to fix this in a cleaner way

**Comment by lauruskyte** [comment]
We are also using poetry, but did anyone solved it without creating a constraints.txt file?

**Comment by Afoucaul** [comment]
> We are also using poetry, but did anyone solved it without creating a constraints.txt file?

I suppose you could do `poetry run pip install "setuptools<72"` then `poetry run pip install --no-build-isolation <your package>` - the constraints.txt file is only a static way of doing it

**Comment by manugarri** [comment]
@lauruskyte for the moment thats all there is. We are waiting to see if the breaking version gets janked.

**Comment by mgierada** [comment]
Oh this sucks so much!

**Comment by adamroyjones** [comment]
Not to be rude, but every comment in this thread (including this one, regrettably) leads to a large number of notifications and noise that drowns out anything useful.

At this point, we should probably wait for a maintainer to provide an update.

**Comment by MarcinWieczorek** [comment]
> This functionality has been deprecated for 5 years

![image](https://media1.tenor.com/m/iEkHX2TqPJ0AAAAC/just-sitting.gif)

**Comment by Jorden28** [comment]
For me the fix for the issue seems to be to run `pip install --upgrade pip-tools pip wheel` before running the command that runs into the `ModuleNotFoundError: No module named 'setuptools.command.test'` error message

**Comment by SarahAndersonValerann** [comment]
For anyone using poetry install in their docker images and are blocked by this, we were able to resolve this by creating a python package dependent on setuptools<72.0.0 and adding that as a dependency to the poetry that doesn't build.
simply adding setuptools<72.0.0 to the blocked poetry didn't work, but seems to if it's the dependent of another package

**Comment by keyz182** [comment]
> > This functionality has been deprecated for 5 years
> 
> ![image](https://camo.githubusercontent.com/f7e69beda5c71fa86b6205aee44faaa1b5a4b8ba22fe5216eaf3c6c528f21204/68747470733a2f2f6d65646961312e74656e6f722e636f6d2f6d2f69456b4858325471504a3041414141432f6a7573742d73697474696e672e676966) [ ![image](https://camo.githubusercontent.com/f7e69beda5c71fa86b6205aee44faaa1b5a4b8ba22fe5216eaf3c6c528f21204/68747470733a2f2f6d65646961312e74656e6f722e636f6d2f6d2f69456b4858325471504a3041414141432f6a7573742d73697474696e672e676966) ](https://camo.githubusercontent.com/f7e69beda5c71fa86b6205aee44faaa1b5a4b8ba22fe5216eaf3c6c528f21204/68747470733a2f2f6d65646961312e74656e6f722e636f6d2f6d2f69456b4858325471504a3041414141432f6a7573742d73697474696e672e676966) [ ](https://camo.githubusercontent.com/f7e69beda5c71fa86b6205aee44faaa1b5a4b8ba22fe5216eaf3c6c528f21204/68747470733a2f2f6d65646961312e74656e6f722e636f6d2f6d2f69456b4858325471504a3041414141432f6a7573742d73697474696e672e676966)


Unfortunately this is what leaving something in a long-term state of deprecation leads to. It becomes de-prioritized. 

If I have code that's been deprecated for an extended period, I prefer to change the deprecation notice to a stronger one including a date when we're ready to remove it, which prompts people to re-prioritize. 

No need to be laying blame anywhere here though, rather look to recommendations to make the process smoother in future. 


In terms of workarounds, `PIP_CONSTRAINTS` alone didn't work for us with poetry, We had to do the following in our pipelines:

```yaml
    - echo "Init Poetry and create requirements"
    - poetry config virtualenvs.in-project true
    - poetry config http-basic.simba $CI_REGISTRY_USER $CI_REGISTRY_PASSWORD
    - poetry export --with dev -o requirements.txt --without-hashes
    - poetry export --with dev -f constraints.txt -o constraints.txt --without-hashes
    - PIP_CONSTRAINT="$(pwd)/constraints.txt" poetry run pip install -r requirements.txt
```

**Comment by manugarri** [comment]
> simply adding setuptools<72.0.0 to the blocked poetry didn't work, but seems to if it's the dependent of another package

Interesting, so if i understood, on a separate package you pinned setuptools to <72.0.0 , and when installing in an environment that has that other package as a dependency, it respects the setuptools constraint?

**Comment by agronholm** [comment]
> For me the fix for the issue seems to be to run `pip install --upgrade pip-tools pip wheel` before running the command that runs into the `ModuleNotFoundError: No module named 'setuptools.command.test'` error message

Why are you installing wheel?

**Comment by jaraco** [comment]
Thanks for the report. I've yanked the 72.0.0 release while I investigate and determine if there's a less invasive way to remove the command.

**Comment by manugarri** [comment]
ok, seems like the version has been yanked, thanks @jaraco  https://github.com/pypa/setuptools/issues/4518#issuecomment-2255928859

**Comment by jaraco** [comment]
Next, let me acknowledge - I did expect this change to have some disruption, but not nearly as much as it had. Sorry for the disruption.

I was anticipating that impacted users would have had visibility to the deprecation warnings and most packages would long ago have migrated and that the handful that remained could be addressed promptly. It seems instead the deprecation warning wasn't placed with enough visibility for those installing affected packages, so the message only served to reduce reliance on the command for running tests, but not necessarily removing the command entirely. This statement is not an excuse, just context.

**Comment by Pragith** [comment]
> For me the fix for the issue seems to be to run `pip install --upgrade pip-tools pip wheel` before running the command that runs into the `ModuleNotFoundError: No module named 'setuptools.command.test'` error message

Thanks @Jorden28, this worked for me.

**Comment by matt-phylum** [comment]
This function can never be removed because removing it renders old package versions non-installable. It's already too late to warn past developers about deprecation.

Python packaging has a huge problem with semantic versioning. The official documentation, [to this day](https://github.com/pypa/setuptools/blob/5e1b3c414779317bc3e105d9bae82ce70c22dbf9/docs/build_meta.rst?plain=1#L59)), recommends to always depend on the latest version of setuptools regardless of major version ([and setuptools is not the only build backend doing this](https://github.com/pypa/hatch/blob/3adae6c0dfd5c20dfe9bf6bae19b44a696c22a43/docs/config/build.md?plain=1#L20)). This means that changes to setuptools can unexpectedly break nearly any sdist package, in this case rendering them non-installable. The change to setuptools version 72.0.0 indicates that it is not compatible with old packages, but the old packages accept any version and don't even specify a single version that is known to be compatible. The builder has no better option than to install 72.0.0 and hope.

The simple way to fix this would be to create a setuptools2 and keep setuptools as a compatible version for packages that use it. However, even if the documentation says to specify a compatible major version, developers may continue not doing it because it's convenient to them at the time, leading to this problem again years later. Maybe setuptools2 can error out if it detects that it is being required without a major version specified to avoid having to create a setuptools3.

To keep the name setuptools, maybe setuptools could detect that it is being required without a major version and replace itself with an older version or compatibility wrapper. It's probably good enough to require the version be set in pyproject.toml and assume anything without pyproject.toml build-backend should use the oldest compatibility mode.

It'd also be possible for the package builder to do this kind of thing if it detects an unconstrained build backend version, but every package builder would need to know about the compatible fallback versions of every package build backend and that sounds like a good way to create inconsistent behavior across builders.

**Comment by keyz182** [comment]
> Next, let me acknowledge - I did expect this change to have some disruption, but not nearly as much as it had. Sorry for the disruption.
> 
> I was anticipating that impacted users would have had visibility to the deprecation warnings and most packages would long ago have migrated and that the handful that remained could be addressed promptly. It seems instead the deprecation warning wasn't placed with enough visibility for those installing affected packages, so the message only served to reduce reliance on the command for running tests, but not necessarily removing the command entirely. This statement is not an excuse, just context.

The good news is I think this has exposed what is maybe not a bug, but an oversight at least on my part - that despite having versions pinned throughout out code base and pipelines, isolated build of packages were not subject to those constraints.

**Comment by faizzed** [comment]
Coming here from https://github.com/AUTOMATIC1111/stable-diffusion-webui

Putting 👇 in requirements.txt fixed it for me.

```
setuptools==69.5.1
```

**Comment by alonme** [comment]
@chrisirhc please update issue to link to
 https://github.com/pypa/setuptools/issues/4519#issuecomment-2255940870 
and
https://github.com/pypa/setuptools/issues/4519#issuecomment-2255961257

**Comment by htelsiz** [comment]
epic morning

**Comment by berkaycamur** [comment]
I just wanted to deploy at Monday instead of Friday and thats what happened

**Comment by ParmarMiral** [comment]
Officially setuptools verion is 71.1.0 again. https://pypi.org/project/setuptools/#history

**Comment by asteel-gsa** [comment]
> The official version of setuptools is now 71.1.0 again. https://pypi.org/project/setuptools/#history 😅

Confirmed local build and github workflow build is working again for our setup after this

**Comment by astrojuanlu** [comment]
> This function can never be removed because removing it renders old package versions non-installable.

That is not correct, workarounds existed before setuptools 72.0.0 got yanked, see https://github.com/pypa/setuptools/issues/4519#issuecomment-2254983472

**Comment by LewisCowlesMotive** [comment]
I just found out now this has been yanked, so I don't know if my fix worked, or I thought I gained understanding, when there was more confusion.

I Think pinning outside of my pyproject.toml (with poetry) fixed because setuptools on the machine is used, rather than the one specified in poetry. Am I wrong about this?

Our original error had python 3.11 and 3.10 in the same traceback, so I `poetry env use 3.11`; no more nonsense about 3.10; but still not fixed.

I pinned in pyproject.toml, and nothing. I was hoping the transitive dependency `pyvcr` would as it's wildcarded, just use whatever is available.

I Also submitted them a PR; but they are being used via `pytest-recording`

I'm uncertain if this is that this installs package dependencies, so following category theory has to exist outside of that; or if the rollback fixed things.

**Comment by CaselIT** [comment]
> Next, let me acknowledge - I did expect this change to have some disruption, but not nearly as much as it had. Sorry for the disruption.
> 
> I was anticipating that impacted users would have had visibility to the deprecation warnings and most packages would long ago have migrated and that the handful that remained could be addressed promptly. It seems instead the deprecation warning wasn't placed with enough visibility for those installing affected packages, so the message only served to reduce reliance on the command for running tests, but not necessarily removing the command entirely. This statement is not an excuse, just context.

Maybe having a daily pipeline that tries to install the top N packages (say top 250 or similar) in pypi using the develop branch would help catching widespread regressions like these?

**Comment by manugarri** [comment]
> > This function can never be removed because removing it renders old package versions non-installable.
> 
> That is not correct, workarounds existed before setuptools 72.0.0 got yanked, see [#4519 (comment)](https://github.com/pypa/setuptools/issues/4519#issuecomment-2254983472)

I disagree Juanlu (hola!) , many build systems are centralised , meaning customers (teams) can provide a pyproject , requirements.txt or similar, and the system takes care of the rest. This means individual python teams do not have the option to specify commands, setup environment variables or other custom changes. 
For example,  in my current company, adding this hotfix would have involved adding quite some hacky code into a non python codebase, just to support a temporary fix.

**Comment by jaraco** [comment]
> Maybe having a daily pipeline that tries to install the top N packages (say top 250 or other reasonable value) in pypi using the develop branch would help catching widespread regressions like these?

See https://github.com/pypa/setuptools/issues/4520#issuecomment-2256130396. There are some tests, but they're not as deep as you might hope. Probably that test could be expanded to the top 100 or top 250.

**Comment by jaraco** [comment]
I'm releasing the hot fix, which restores a small amount of the previous code to limit the breakage to "running setup.py test" but doesn't break building a package that is importing the command, adding a deprecation warning for that case. It's releasing as v72.1.0. I'll be keeping a close eye on this issue and the issue tracker in case that change doesn't adequately limit the disruption.

**Comment by davidaurelio** [comment]
Does that mean that building old projects relying on this will break again? If that were the case, we'd end up with a broken CI pipeline once more. Since we rely on `uv` we have no means of controlling which version of setuptools will be pulled in when building projects.

In our case, one of the breaking packages was `ffmpy`: https://github.com/Ch00k/ffmpy/blob/master/setup.py#L2

Please hold off with a new release at least until `uv` provides a workaround.

**Comment by agronholm** [comment]
> Does that mean that building old projects relying on this will break again? If that were the case, we'd end up with a broken CI pipeline once more. Since we rely on `uv` we have no means of controlling which version of setuptools will be pulled in when building projects.
> 
> In our case, one of the breaking packages was `ffmpy`: https://github.com/Ch00k/ffmpy/blob/master/setup.py#L2
> 
> Please hold off with a new release at least until `uv` provides a workaround.

No, IIRC only explicitly running `python setup.py test` would give an error, but merely having a `setup.py` referencing the test command would not.

**Comment by jaraco** [comment]
Given the lack of noise or reports since the release of 72.1.0, I believe the fix addressed the vast majority of use-cases affected by 72.0.0. My hope is that any remaining issues are isolated and readily addressable, but feel free to follow up here or in a new issue.

I want to thank everybody who was impacted by this for remaining civil, for your efforts in finding and documenting workarounds, and for providing constructive feedback on a change that was unnecessarily hostile to the ecosystem.

**Comment by nykcrystal** [comment]
In my case , I deupgrade my setuptools to version 71.1 solved the problem as well

**Comment by ziga7631** [comment]
In my case, the following 2 options work using Poetry:
1. First installing the problematic module and then the remaining ones
![image](https://github.com/user-attachments/assets/cbf89fab-ea59-4510-b6e2-e6bb2af1c408)
2. Specifying build-system requirement for Setuptools
![image](https://github.com/user-attachments/assets/1aca9782-8d62-4838-a60e-dcd5d4dda035)

Should the second option also be working, or is this only some coincidence. It is by far the easiest solution using Poetry.

**Comment by justinclift** [comment]
@ziga7631 The problem release (`72.0.0`) has been yanked.  So in theory things should be ok now anyway.  Just make sure your system isn't using `72.0.0` specifically.

There's also a newer `72.1.0` release with a fix in it as well.
